### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDeps** [serengil/deepface#511]
Hi, In **deepface**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using
```
numpy>=1.14.0
pandas>=0.23.4
gdown>=3.10.1
tqdm>=4.30.0
Pillow>=5.2.0
opencv-python>=4.5.5.64
opencv-contrib-python>=4.3.0.36
tensorflow>=1.9.0
keras>=2.2.0
Flask>=1.1.2
mtcnn>=0.1.0
lightgbm>=2.3.1
dlib>=19.20.0
retina-face>=0.0.1
mediapipe>=0.8.7.3
fire>=0.4.0
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **keras** can be changed to *>=0.2.0,<=2.1.6*.
The version constraint of dependency **keras** can be changed to *>=2.2.3,<=2.3.1*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

    
<details>
<summary>The calling methods from the keras</summary>
<pre>
image.img_to_array
</pre>
</details>

    
<details>
<summary>The calling methods from the  all methods</summary>
<pre>
training.Model
cap.read
embeddings.append
test_cases
df_train.drop
OpenFace.loadModel
isinstance
emotion_df.sort_values
model.predict
detected_faces.append
url.split
tensorflow.keras.layers.BatchNormalization
tf.get_default_graph
Boosting.build_gbm
FbDeepFace.loadModel
pbar.set_description
InceptionResNetV2
tensorflow.keras.layers.ZeroPadding2D
emotion_df.sort_values.reset_index
max
enumerate
np.float64
cv2.addWeighted
np.argmax
jsonify
os.path.isdir
models.predict
x.tf.nn.lrn.Lambda
parser.parse_args
keras.layers.Flatten
detections_df.astype
name.tensorflow.keras.layers.PReLU
Convolution2D
functions.initialize_input
DlibResNet
df.to_csv
cv2.waitKey
functions.find_input_shape
OpenCvWrapper.align_face
np.sqrt
postprocess.alignment_procedure
feature.df.decision.df.plot
np.matmul
AveragePooling2D
np.array
feature_names.append
model.copy
_Layer
img2_path.any
distance.findEuclideanDistance
lgb.Booster
tensorflow.keras.layers.Input
fig.add_subplot
pickle.load
detect_faces
open
Concatenate
tf.get_logger
VGGFace.loadModel
type
employee.split
cv2.imdecode
face_detector
dst.findThreshold
Dropout
np.sum
face_detector.detect_faces
race_predictions.sum
models.get
mp_face_detection.FaceDetection
len
distance_vector.sum
df_test.drop
img_aligned.astype
model_pbar.set_description
keras.layers.Dense
functions.preprocess_face
cv2.line
model_names.append
actions.append
np.square
LocallyConnected2D
FaceDetector.detect_face
np.transpose
keras.layers.BatchNormalization
bool
DeepFace.find
negatives.positives.pd.concat.reset_index
pd.read_csv
k.j.df.sort_values.reset_index
os.makedirs
list
Activation
gender_model.load_weights
negative.append
str
name.bn_axis.tensorflow.keras.layers.BatchNormalization
dlib.shape_predictor
race_model.load_weights
DeepFace.detectFace
base_img.copy
uri.split
functions.get_deepface_home
d.bottom
roc_curve
Dense
cv2.VideoCapture
cv2.destroyAllWindows
Boosting.validate_model
plt.colorbar
x.Lambda
Lambda
img.std
target_name.df.drop.columns.tolist
verified_labels.append
detect_face
x.K.l2_normalize.Lambda
stack_fn
confusion_matrix
prediction_classes.append
tf_version.split
emotion_predictions.sum
concatenate
pickle.dump
tensorflow.keras.layers.PReLU
Model
represent
model.compute_face_descriptor
detected_faces.copy
d.left
img_aligned.max
cv2.cvtColor
raw_img.copy
gdown.download
img.model.predict.tolist
Facenet.InceptionResNetV2
name.tensorflow.keras.layers.Add
requests.get
DeepFace.stream
idendities.items
Path.home
MTCNN
ResNet34
model
model.items
image.astype
Age.findApparentAge
pd.concat
baseModel
graph.as_default
get_deepface_home
lgb.plot_importance
FaceDetector.alignment_procedure
tf.nn.lrn
negatives.sample
lgb.train
globals
instances.append
time.time
OpenCvWrapper.build_cascade
app.route
K.l2_normalize
cv2.dnn.readNetFromCaffe
zipfile.read
employees.append
Sequential
base_threshold.model_name.thresholds.get.get
resp_obj.append
tuple
cv2.CascadeClassifier
ZeroPadding2D
os.walk
VGGFace.baseModel
cm.ravel
RetinaFace.detect_faces
zipfile.ZipFile
df.values.tolist
df.copy
metrics.append
K.sqrt
functions.normalize_input
SFaceModel
cap.release
base64.b64decode
dst.l2_normalize
itertools.product
cv.FaceRecognizerSF.create
employee_name.split
min
parser.add_argument
cv2.imread
df.head
os.path.exists
BatchNormalization
np.multiply
detector.detectMultiScale
base_model.add
app.run
emotion_df.iterrows
train_test_split
self.model.feature
path.exists
MaxPooling2D
instance.append
resp.append
gender_model.predict
representWrapper
opencv_home.split
plt.show
functions.initialize_folder
file_name.replace.lower
img1_path.copy
print
np.pad
keras.layers.Dropout
age_model.predict
embedding.append
eye_detector.detectMultiScale
resp_objects.append
cv2.rectangle
verifyWrapper
Image.open
df.sort_values.reset_index
dlib.face_recognition_model_v1
model.add
cv2.resize
get_opencv_path
uuid.uuid4
df.apply
plt.figure
FaceDetector.build_model
found_models.append
img1_graph.append
sp
positive.append
abs
backends.get
stack1
roc_auc_score
name.stride.filters.tensorflow.keras.layers.Conv2D
name.tensorflow.keras.layers.ZeroPadding2D
Flatten
age_model.load_weights
plt.plot
model.load_weights
fire.Fire
employee_name.split.replace
add
tensorflow.keras.layers.Conv2D
build_cascade
loadBase64Img
ord
img2_graph.append
up.K.int_shape.scaling.Lambda
idendities.values
zip_ref.extractall
re.sub
DeepFace.analyze
block1
mood_items.append
lgb.Dataset
cv2.drawContours
cv2.putText
freeze_img.copy
cv2.dnn.blobFromImage
tf.__version__.split
f.close
face_detector.process
evaluate
tf.get_logger.setLevel
negatives.append
columns.append
setuptools.find_packages
tqdm
dst.findEuclideanDistance
dlib.get_frontal_face_detector
boosted_tree.predict
round
file.lower
dlib.get_face_chip
int
set
demography.get
df.decision.value_counts
name.filters.tensorflow.keras.layers.Conv2D
plt.text
dst.findCosineDistance
pd.DataFrame
representations.append
face_detector_obj.keys
np.fromstring
argparse.ArgumentParser
df.iterrows
img.mean
load_image
tensorflow.keras.layers.Add
req.keys
ValueError
d.right
d.top
analyzeWrapper
os.path.isfile
img.startswith
scores.append
format
keras.models.Model
Flask
realtime.analysis
build_model
img1.split
bz2.BZ2File
plt.imshow
DeepFace.build_model
name.stride.kernel_size.filters.tensorflow.keras.layers.Conv2D
Facenet.loadModel
detections_df.iterrows
Image.fromarray
emotion_model.predict
models.keys
newfilepath.open.write
DeepFace.verify
DeepFace.represent
GlobalAveragePooling2D
K.int_shape
plt.axis
file_name.replace
DlibMetaData
Boosting.loadModel
request.get_json
x.K.sqrt.Lambda
distances.append
dimension.Dense
classes.Convolution2D
np.expand_dims
Conv2D
Input
gbm.predict
align_face
setuptools.setup
face_detector.forward
df.sort_values
Add
np.arccos
cv2.imshow
image.img_to_array
positives.append
base_model.load_weights
FaceDetector.detect_faces
range
mood_item.append
img.rotate
thresholds.get
warnings.filterwarnings
sorted
ensemble_features.append
face_detector.setInput
img.copy
df.drop
tqdm.pandas
fh.read
RetinaFace.build_model
os.getenv
model_obj.keys
metric_names.append
gbm.save_model
img2.split
distance_graph.append
df_base.copy
</pre>
</details>

    
@sugyan 
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.

**Comment by serengil** [comment]
any contributions are welcome!

i am currently using the minimum versions mentioned in requirements.txt. i mean that i am using exact keras 2.2.0 if it is menioned >=2.2.0.

so i am closing the issue because this is not directly a bug.
